### PR TITLE
Cleanup in preparation for Font Styles

### PR DESCRIPTION
**Pull request by charlesroddie**
Initial simplification in preparation for adding Font Styles
- DoNothingFontChanger was removed was only used in tests.

For subsequent PR:
- UnicodeFontChanger uses "mathematical bold/italic" ranges to use a single font for normal/bold/italic. To preserve existing functionality without deploying additional fonts, this logic needs to be kept when using standard fonts.
- Allow users to add bold and italic fonts, for use in TextPainter. This PR will not address MathAtom fonts. If these fonts are present, they will be used instead of UnicodeFontChanger for rendering bold/italic fonts.

### CODE REVIEW

**Comment by Happypig375** [comment]
>ancient LaTeX approach of using "mathematical bold/italic" ranges

Actually, LaTeX predates Unicode so it looks up font files of different styles. WpfMath does this: they have different fonts for the same codepoint that contain different styles. However, iosMath uses Latin Modern Math by default which combines all these fonts together into the mathematical glyph ranges, so we can use Unicode from top to bottom.

**Comment by charlesroddie** [comment]
> Actually, LaTeX predates Unicode so it looks up font files of different styles. WpfMath does this: they have different fonts for the same codepoint that contain different styles. However, iosMath uses Latin Modern Math by default which combines all these fonts together into the mathematical glyph ranges, so we can use Unicode from top to bottom.

Interesting. WpfMath then ends up with a more convenient approach with fewer special cases but I think this PR can go ahead without changing that. We'll just need to preserve some degree of "font changing" for TextAtoms.

**Comment by charlesroddie** [comment]
Can you understand the test failures, or do you remember why there was a DoNothingFontChanger in the tests (before the changes in the cleanup so far) @Happypig375 ?

It would be nice to remove test failures before pushing the real work. I can't see any stated meaning of the tests so cannot evaluate. Presumably the tests were not testing actual behavior (since they required DoNothingFontChanger to pass) and so removing DoNothingFontChanger is a matter of hygiene so that the tests test real behaviour.

**Comment by Happypig375** [comment]
>Can you understand the test failures, or do you remember why there was a DoNothingFontChanger in the tests (before the changes in the cleanup so far) @Happypig375 ?

It's just a mock for UnicodeFontChanger in TypesettingContext. It's a leftover from @verybadcat.
>I can't see any stated meaning of the tests so cannot evaluate.

Yes, there are some tests that test the mocks themselves.

**Comment by Happypig375** [comment]
The test failures are related to using UnicodeFontChanger instead of DoNothingFontChanger. Since the text now have characters outside of the BMP, the difference in UTF16 length leads to different IndexRanges. Might have to check iosMath on its behaviour.

**Comment by Happypig375** [comment]
This anomaly can be traced all the way back here:
https://github.com/verybadcat/CSharpMath/blob/50f10f5263687fe98df6249381a73d087ac979bf/CSharpMath/Display/LinesAndRuns/TextLineDisplays.cs#L22-L29
As far as I understand, ranges represent locations of atoms before finalization. However, here, it is not used in this way. Hmmm... Is this unintentional?

**Comment by Happypig375** [comment]
Definitely not correct. CSharpMath.Editor uses Ranges as input to constructing MathListIndexes. Ranges must represent locations of atoms and using text length instead is a logic error.

**Comment by charlesroddie** [comment]
If the tests are not right it's best to delete them rather than try to correct them. Tests on intermediate output (in between latex and atom), even if they were documented and functioning, aren't necessary.

Sent from Nine<http://www.9folders.com/>
________________________________

**Comment by Happypig375** [comment]
You mean UnicodeFontChangerTests?

**Comment by charlesroddie** [comment]
UnicodeFontChanger/Range tests.

**Comment by Happypig375** [comment]
Range tests are important since they are part of the atom output. Not part of the intermediate steps in creating atoms.

**Comment by Happypig375** [comment]
Alright, all tests now pass.

**Comment by Happypig375** [comment]
@charlesroddie I've changed UnicodeFontChangerTests so that they also test LaTeX->Display conversion instead of only testing UnicodeFontChanger.

File: CSharpMath.CoreTests/FontChangingTests.cs [hunk 1]
```diff
@@ -0,0 +1,155 @@
+namespace CSharpMath.CoreTests {
+  using Xunit;
+  using Atom;
+  public class FontChangingTests {
+    // Tests are ordered by numeric value of the Atom.FontStlye enumeration
+    void Test(string input, string output, string command) {
```
Comment (charlesroddie) [inline CSharpMath.CoreTests/FontChangingTests.cs:6]: Needs a name + documentation

File: CSharpMath.CoreTests/TypesetterTests.cs [hunk 2]
```diff
@@ -33,15 +33,15 @@ public class TypesetterTests {
     void TestOuter(string latex, int rangeMax, double ascent, double descent, double width,
         params System.Action<IDisplay<TFont, TGlyph>>[] inspectors) =>
       TestList(rangeMax, ascent, descent, width, 0, 0, LinePosition.Regular, Range.UndefinedInt, inspectors)
-      (Typesetter.CreateLine(LaTeXParserTest.ParseLaTeX(latex), _font, _context, LineStyle.Display));
+      (ParseLaTeXToDisplay(latex));
 
-    [Theory, InlineData("x"), InlineData("2")]
-    public void TestSimpleVariable(string latex) =>
+    [Theory, InlineData("x", "𝑥"), InlineData("2", "2"), InlineData(".", "."), InlineData("𝑥", "𝑥")]
+    public void TestSingleCharacter(string latex, string text) =>
```
Comment (charlesroddie) [inline CSharpMath.CoreTests/TypesetterTests.cs:39]: Is it known what this test is for? If not, we should delete. If so, we should rename/document.

File: CSharpMath.CoreTests/TypesetterTests.cs [hunk 3]
```diff
@@ -68,15 +68,15 @@ public class TypesetterTests {
         });
     [Theory]
     [InlineData("%\n1234", "1234")]
-    [InlineData("12.b% comment ", "12.b")]
+    [InlineData("12.b% comment ", "12.𝑏")]
     [InlineData("|`% \\notacommand \u2028@/", "|`@/")]
-    public void TestComments(string latex, string resultText) =>
+    public void TestComments(string latex, string text) =>
```
Comment (charlesroddie) [inline CSharpMath.CoreTests/TypesetterTests.cs:73]: Is it known what this test is for? If not, we should delete. If so, we should rename/document.

File: CSharpMath.CoreTests/_Helpers/FrontEnd/TestGlyphNameProvider.cs [hunk 4]
```diff
@@ -1,8 +1,33 @@
 namespace CSharpMath.CoreTests.FrontEnd {
-  class TestGlyphNameProvider : Display.FrontEnd.IGlyphNameProvider<char> {
+  using System.Linq;
+  using TGlyph = System.Text.Rune;
+  class TestGlyphNameProvider : Display.FrontEnd.IGlyphNameProvider<TGlyph> {
```
Comment (charlesroddie) [inline CSharpMath.CoreTests/_Helpers/FrontEnd/TestGlyphNameProvider.cs:4]: Do we know what this is?

**Review by charlesroddie — COMMENTED** [review]
Thanks for getting the tests to pass. There are some additional tests included here that need documentation.

**Comment by charlesroddie** [comment]
I think we should merge this as a cleanup and then I'll do a separate PR with the real work.

File: CSharpMath.CoreTests/TypesetterTests.cs (hunk 2 above)
Comment (Happypig375) [reply CSharpMath.CoreTests/TypesetterTests.cs:39]: Makes sure that a single character of type Variable/Number/Punctuation/Ordinary(surrogates) measures as the same one-character size.

File: CSharpMath.CoreTests/TypesetterTests.cs (hunk 3 above)
Comment (Happypig375) [reply CSharpMath.CoreTests/TypesetterTests.cs:73]: Makes sure that comments are ignored correctly when generating displays

File: CSharpMath.CoreTests/_Helpers/FrontEnd/TestGlyphNameProvider.cs (hunk 4 above)
Comment (Happypig375) [reply CSharpMath.CoreTests/_Helpers/FrontEnd/TestGlyphNameProvider.cs:4]: Looks up a name in latinmodern-math.json

File: CSharpMath.CoreTests/FontChangingTests.cs (hunk 1 above)
Comment (Happypig375) [reply CSharpMath.CoreTests/FontChangingTests.cs:6]: I think the code is self-explanatory?

File: CSharpMath.CoreTests/TypesetterTests.cs (hunk 3 above)
Comment (charlesroddie) [reply CSharpMath.CoreTests/TypesetterTests.cs:73]: OK so rename to CommentsAreIgnoredWhenGeneratingDisplays or add xml comment `Tests that comments are ignored correctly when generating displays

Comment (Happypig375) [reply CSharpMath.CoreTests/TypesetterTests.cs:73]: I think just inserting the word "Ignore" will be sufficient here.

File: CSharpMath.CoreTests/TypesetterTests.cs [hunk 5]
```diff
@@ -33,15 +33,15 @@ public class TypesetterTests {
...
+      (ParseLaTeXToDisplay(latex));
 
-    [Theory, InlineData("x"), InlineData("2")]
-    public void TestSimpleVariable(string latex) =>
+    [Theory, InlineData("x", "𝑥"), InlineData("2", "2"), InlineData(".", "."), InlineData("𝑥", "𝑥")]
+    public void TestSingleCharacter(string latex, string text) =>
       TestOuter(latex, 1, 14, 4, 10,
         d => {
           var line = Assert.IsType<TextLineDisplay<TFont, TGlyph>>(d);
           Assert.Single(line.Atoms); // have to think about these; doesn't really work atm
```
Comment (FoggyFinder) [inline CSharpMath.CoreTests/TypesetterTests.cs:43]: is this comment outdated?

Comment (Happypig375) [reply CSharpMath.CoreTests/TypesetterTests.cs:43]: @verybadcat wrote it but I think `Atoms` work fine.

**Comment by Happypig375** [comment]
@charlesroddie 2 outstanding reviews should be resolved for this to be mergable

**Comment by charlesroddie** [comment]
@Happypig375 Are the two things to be resolved the questions I asked about the tests? I don't understand these well enough to add documentation.

File: CSharpMath.CoreTests/TypesetterTests.cs (hunk 2 above)
Comment (charlesroddie) [reply CSharpMath.CoreTests/TypesetterTests.cs:43]: I copied this comment in but don't understand the terms involved. You'll have to judge whether this is resolved.

File: CSharpMath.CoreTests/_Helpers/FrontEnd/TestGlyphNameProvider.cs (hunk 4 above)
Comment (charlesroddie) [reply CSharpMath.CoreTests/_Helpers/FrontEnd/TestGlyphNameProvider.cs:5]: Again, copied in your comment. To be in line with the description, should TestGlyphNameProvider be renamed to LatinModernGlyphNameProvider?

Comment (Happypig375) [reply CSharpMath.CoreTests/_Helpers/FrontEnd/TestGlyphNameProvider.cs:5]: This is strictly for tests to approximate glyph lookup. The correct and comprehensive implementation of this is in CSharpMath.Apple.

File: CSharpMath.CoreTests/TypesetterTests.cs (hunk 2 above)
Comment (Happypig375) [reply CSharpMath.CoreTests/TypesetterTests.cs]: Clarified

**Comment by charlesroddie** [comment]
Can this be merged?

**Comment by Happypig375** [comment]
There are 2 unresolved conversations

**Comment by charlesroddie** [comment]
You understand the remaining conversation so can you mark if you consider it resolved?

**Review by Happypig375 — APPROVED** [review]

**Review by FoggyFinder — APPROVED** [review]

**Comment by Atlantis433** [comment]
Charlie and whomever pay attention to the master here and I am here to tell you both that the house is due to be cleaned of the filth and chaos that you have done. You have many incorrect coding and you're not even close to be verified. The mit license doesn't belong to you at all it's mine. Your hatred is a terrible thing to have and I didn't do anything wrong for you to do this.

**Pull request merged**